### ISSUE TITLE
Update the header format for CSV

### ISSUE THREAD

**Issue by abbyad** [medic/medic-conf#47]
The current header naming structure is functional but confusing, especially that there are 3 different formats that work differently. We should update the header format to make it easy to use for tech leads and partners. Below is a proposed format that has a common format for plain headers, those with external references matching a field, and those matching to a row number.

## CURRENT STRUCTURES:
1. No external reference, with optional conversion to Type:
[ `{{TYPE}}:` ]`{{TARGET}}`

1. Reference to a CSV row:
`csv:{{doc_type}}:` [ `id` | `doc` | `.source_field` ] `>target_field_name`
  _Note: unclear if type conversion is currently supported with this format_

1. Match to another doc:
`match={{field_in_other_doc}}:field1=value1&...&fieldN=fieldN:` [ `id` | `doc` | `.source_field` ] `>target_field_name`
  _Note: unclear if type conversion is currently supported with this format_

## PROPOSED STRUCTURE:
[ `{{TYPE}}:` ] [ `{{SOURCE}}:` ] `{{TARGET}}`

`TYPE` and `SOURCE` are optional. Without them we have the following:

`{{TARGET}}`
No type conversion or external reference

`{{TYPE}}:{{TARGET}}`
No external reference, set the type for conversion of data in the column. 

`{{SOURCE}}:{{TARGET}}`
No conversion done, and the target gets the result of the reference as a string or object

It should be easy enough to differentiate the two 2-part formats since TYPE  is one of 6 possible types that don't overlap with the SOURCE types.

### TYPE
The type of data for this field, saved as corresponding JSON type:
- `string`      default
- `number`
- `boolean`
- `object`
- `date`           date in ISO8601 format
- `timestamp`   date in MS since epoch. Useful for reported_date

### SOURCE
The docs in which to look for the reference:
`{{doc_type}}`[`?match={{field_name}}&{{field_X}}={{value_x}}&{{field_Y}}={{value_y}}`]
- The `doc_type` is the `type` of the doc you are referencing: [`data_record`, `person`, `clinic`, `health_center`, `district_hospital`]. 
- With `match` the row's value is matched to the value of the specified column in the source docs. If `match` is missing the column's value (a number) will be used to find the corresponding row number in the source csv. See _Questions_ section about whether we need to support row number matching.
- Further narrowing can be done with field/value pairs.
- If the source is empty that is ok, it is just not an external reference!

### TARGET
Assigning the data to a field name in the resulting JSON:

[`{{source_field}}>`]`{{target_field}}` where `source_field` could be one of the following:
- `.field_name` for any field on the object
- `doc` for the whole doc (might not be needed at all, see _Questions_ below)
- `id` for the `_id` value (might not be needed at all, see _Questions_ below)

Some examples:
- `target_property_name`
- `doc>target_property_name`
- `id>target_property_name`
- `._id>target_property_name`
- `._id>target_property_name._id`
- `._name>target_property_name`

## EXAMPLES
Here are some column headers in 1.6.12 vs proposed structure

### Without reference
`sex` --> `::sex` 
  or `sex` (no change!)

`timestamp:reported_date`  --> `timestamp::reported_date`
  or `timestamp:reported_date` (no change!)

### Reference by row number
`csv:F:id>target_property_name` --> `string:F:._id>target_property_name._id` 
  or `F:._id>target_property_name._id`

`csv:F:doc>target_property_name` --> `object:F:target_property_name` 
  or `F:target_property_name`

`csv:F:.some_field>target_property`
  -->`string:F:.some_field>target_property`
  or `F:.some_field>target_property`

### Reference with match
`match=P:X=a&Y=y:id>target_property_name` 
  --> `string:F?match=P&X=a&Y=y:._id>target_property_name._id`
  or `F?match=P&X=a&Y=y:._id>target_property_name._id`

`match=P:X=a&Y=y:doc>target_property_name`
  --> `object:F?match=P&X=a&Y=y:target_property_name`
  or `F?match=P&X=a&Y=y:target_property_name`

`match=P:X=a&Y=y:.some_field>target_property_name`
  --> `string:F?match=P&X=a&Y=y:.some_field>target_property_name`
 or `F?match=P&X=a&Y=y:.some_field>target_property_name`

`match=external_id:type=clinic:id>parent` 
  --> `string:clinic?match=external_id:._id>parent._id`
  or `clinic?match=external_id:._id>parent._id`

`match=external_id:type=clinic:doc>parent`
  --> `object:clinic?match=external_id:parent`
  or `clinic?match=external_id:parent`

`match=external_id:type=person:._id>fields.patient_id` 
  --> `string:person?match=external_id:._id>parent_id`
  or `person?match=external_id:._id>parent_id`

`match=external_id:type=person:doc>contact` 
  --> `object:person?match=external_id:contact`
  or `person?match=external_id:contact`

## QUESTIONS
- Is it necessary to have the `doc` and `id` keywords? Seems like we can eliminate them and still get the same result by assigning whatever the value is, either an `object` or `string`. Without it it is pretty clear what is going on and avoids needing to learn/understand the keywords.
- Do we need reference by row number? If there is no use case for this perhaps we can remove that to reduce complexity.
- Should we consider allowing type=place for looking at any place type? If so, differentiation between place types can still be done with the field matching eg type=clinic

**Comment by alxndrsn** [comment]
> Do we need reference by row number? If there is no use case for this perhaps we can remove that to reduce complexity.

It doesn't work with `doc_type` as given in this example anyway, so if we can eliminate it, we'll save some discussion.  On the other hand, if reference-by-row is not available, some kind of extra ID field will have to be included in every CSV.

**Comment by abbyad** [comment]
> Where has F come from in the match examples?

These were based off of the examples in the [README](https://github.com/medic/medic-conf/tree/e04b4f6ebf601d0e9e3f8dfa843246c0de41726e#reference-to-a-row-of-csv)

> It doesn't work with doc_type as given in this example anyway, so if we can eliminate it, we'll save some discussion. On the other hand, if reference-by-row is not available, some kind of extra ID field will have to be included in every CSV.

In our latest import we created unique `external_id` for matching purposes only. These ended up being stored with each imported doc but didn't have to. Having a way to reference another doc without storing the value is useful, yet referring to CSV lines would not have been manageable. 

Perhaps we can use a specific header (eg `_match_key`) that is used solely for matching, and not uploaded. This would simplify things by removing the row referencing, and avoid uploading a field that has outlived it's purpose.

> Including source_field in the TARGET section doesn't seem ideal. Could it be appended to the doc_type in SOURCE?

I agree, that makes more sense! Looking back I think I had proposed it the other way since that was how it was, which I could justify because SOURCE deals with the whole row/doc, and TARGET deals with individual fields. Having the `source_field` with doc in the SOURCE would be cleaner and eliminate the `>` notation.
